internal/pkg/cli/deploy/env.go

```diff
@@ -100,11 +139,154 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
+func (d *envDeployer) Validate(ctx context.Context, mft *manifest.Environment, output io.Writer) error {
+	isManagedCDNEnabled := mft.CDNEnabled() && !mft.HasImportedPublicALBCerts() && d.app.Domain != ""
+	if isManagedCDNEnabled {
+		// With managed domain, if the customer isn't using `alias` the A-records are inserted in the service stack as each service domain is unique.
+		// However, when clients enable CloudFront, they would need to update all their existing records to now point to the distribution.
+		// Hence, we force users to use `alias` and let the records be written in the environment stack instead.
+		if err := d.envDescriber.ValidateCFServiceDomainAliases(); err != nil {
+			return err
+		}
+	}
+
+	if mft.CDNEnabled() && mft.CDNDoesTLSTermination() && (mft.HasImportedPublicALBCerts() || d.app.Domain != "") {
+		err := d.validateALBWorkloadsDontRedirect(ctx)
+		var redirErr *errEnvHasPublicServicesWithRedirect
+		switch {
+		case errors.As(err, &redirErr) && mft.ALBIngressRestrictedToCDN():
+			return err
+		case errors.As(err, &redirErr):
+			fmt.Fprintln(output, redirErr.Warning())
+		case err != nil:
+			return fmt.Errorf("can't enable TLS termination on CDN: %w", err)
+		}
+	}
+
+	return nil
+}
+
+type errEnvHasPublicServicesWithRedirect struct {
+	services []string
+}
+
+func (e *errEnvHasPublicServicesWithRedirect) Error() string {
+	n := len(e.services)
+	quoted := make([]string, len(e.services))
+	for i := range e.services {
+		quoted[i] = strconv.Quote(e.services[i])
+	}
+
+	return fmt.Sprintf(`%s %s %s HTTP traffic to HTTPS.
+%s
+To fix this, set the following field in %s manifest:
+%s
+and run %s.`,
+		english.PluralWord(n, "Service", "Services"),
+		english.OxfordWordSeries(quoted, "and"),
+		english.PluralWord(n, "redirects", "redirect"),
+		color.Emphasize(english.PluralWord(n, "This service", "These services")+" will not be reachable through the CDN."),
+		english.PluralWord(n, "its", "each"),
+		color.HighlightCodeBlock("http:\n  redirect_to_https: true"),
+		color.HighlightCode("copilot svc deploy"),
+	)
+}
+
+func (e *errEnvHasPublicServicesWithRedirect) Warning() string {
+	return fmt.Sprintf(`%s
+If you'd like to use %s without a CDN, ensure %s A record is pointed to the ALB.`,
+		e.Error(),
+		english.PluralWord(len(e.services), "this service", "these services"),
+		english.PluralWord(len(e.services), "its", "each service's"),
+	)
+}
+
+// validateALBWorkloadsDontRedirect verifies that none of the public ALB Workloads
+// in this environment have a redirect in their HTTPWithDomain listener.
+// If any services redirect, an error is returned.
+func (d *envDeployer) validateALBWorkloadsDontRedirect(ctx context.Context) error {
+	params, err := d.envDescriber.Params()
+	if err != nil {
+		return fmt.Errorf("get env params: %w", err)
+	}
+
+	services := strings.Split(params[cfnstack.EnvParamALBWorkloadsKey], ",")
+	g, ctx := errgroup.WithContext(ctx)
```

Comment:
true.............................................i _wish_ i had a good reason to keep it in `svc_deploy.go`, since that's where the whole chain "starts"......but i guess i don't ¯\_(ツ)_/¯